versioned_docs/version-2020fa/lecture8.md

```diff
@@ -3,4 +3,561 @@ id: lecture8
+you'll get the `any` type.
+
+```typescript
+// anyResponse: AxiosResponse<any>
+const anyResponse = await axios.get('url');
+// anyData: any
+const anyData = anyResponse.data;
+
+type DataType = ...;
+// typedResponse: AxiosResponse<DataType>
+const typedResponse = await axios.get<DataType>('url');
+// typedData: DataType
+const typedData = typedResponse.data;
+```
+
+## CORS workaround: proxy
+
+When you are testing your backend and frontend together on localhost, you will
+come across some sort of **CORS policy error** when you try to call the backend
+from the frontend.
+
+One way to get around this is to add this line to your **frontend**'s
+`package.json`:
+
+```json
+  "proxy": "http://localhost:8080",
+```
+
+where `8080` is the the port of your **backend** (change it accordingly).
+
+After adding the proxy config, remove the base URL from your requests:
+
+```typescript
+// do this
+const posts = await axios.get('/getAllPosts');
+
+// instead of this
+const posts = await axios.get('http://localhost:8080/getAllPosts');
+```
+
+## Summary
+
+You keep track of your data in component state. Your app displays whatever data
+you have.
+
+You update your data by calling an endpoint within `useEffect` and setting your
+data to the response that you get back.
+
+You can call endpoints using `fetch()` or `axios` and handle the responses
+asynchronously.
+
+## Demo Code
+
+### Backend
+
+```typescript title="index.ts (backend)"
+import admin from 'firebase-admin';
+import express from 'express';
+
+const serviceAccount = require('./service-account.json');
+
+admin.initializeApp({
+  credential: admin.credential.cert(serviceAccount),
+  databaseURL: '[DATABASE_URL]',
+});
+
+const db = admin.firestore();
+
+const app = express();
+const port = 8080;
+app.use(express.json());
+
+type Song = {
+  name: string;
+  artist: string;
+  rating: number;
+};
+
+type SongWithID = Song & {
+  id: string;
+};
+
+const songsCollection = db.collection('songs');
+
+app.get('/getSongs', async (req, res) => {
+  const songs = await songsCollection.get();
+  res.json(
+    songs.docs.map(
+      (doc): SongWithID => {
+        const song = doc.data() as Song;
+        return { ...song, id: doc.id };
+      }
+    )
+  );
+});
+
+app.post('/createSong', async (req, res) => {
+  const newSong: Song = req.body;
+  const addedSong = await songsCollection.add(newSong);
+  res.send(addedSong.id);
+});
+
+app.post('/updateRating', async (req, res) => {
+  const { id, rating } = req.query;
+  await songsCollection.doc(id as string).update({ rating });
+  res.send('Song rating updated!');
+});
+
+app.listen(port, () => console.log(`Example app listening on port ${port}!`));
+```
+
+```tsx title="SongList.tsx (frontend)"
+import React, { useState, useEffect } from 'react';
+import { Song } from './Song';
+import { SongAdder } from './SongAdder';
+import axios from 'axios';
+
+type Song = {
+  name: string;
+  artist: string;
+  rating: number;
```

Comment:
let's add readonly to all the props field to be consistent with previous notes

Reply:
These aren't exactly props, but I think readonly makes sense still.

What do you think about making fields readonly in the backend code?

Reply:
no strong opinion on backend code. frontend code is more important because it prevents students from changing array and wondering why react doesn't rerender.